app/models/agents/gap_detector_agent.rb

```diff
@@ -42,26 +46,81 @@ def working?
 
     def receive(incoming_events)
       incoming_events.sort_by(&:created_at).each do |event|
-        memory['newest_event_created_at'] ||= 0
-
         if !interpolated['value_path'].present? || Utils.value_at(event.payload, interpolated['value_path']).present?
-          if event.created_at.to_i > memory['newest_event_created_at']
-            memory['newest_event_created_at'] = event.created_at.to_i
-            memory.delete('alerted_at')
+          if stream_path
+            @stream = Utils.value_at(event.payload, stream_path)
+            next unless @stream
+          end
+          if event.created_at.to_i > newest_event_created_at
+            self.newest_event_created_at = event.created_at.to_i
+            clear_alerted_at
           end
         end
       end
     end
 
     def check
       window = interpolated['window_duration_in_days'].to_f.days.ago
-      if memory['newest_event_created_at'].present? && Time.at(memory['newest_event_created_at']) < window
-        unless memory['alerted_at']
-          memory['alerted_at'] = Time.now.to_i
-          create_event payload: { message: interpolated['message'],
-                                  gap_started_at: memory['newest_event_created_at'] }
+      if stream_path
+        memory.each_key do |stream|
+          next unless memory[stream].is_a? Hash
+          next unless memory[stream].has_key? 'newest_event_created_at'
+          @stream = stream
+          check_stream(window)
```

Comment:
I thought about that, but it needs to make it down to event_storage, which means passing into all of the newest_event_* and alerted_at*.  I'm not crazy about this solution either, but seemed to be less messy, especially given that stream is optional.